### ISSUE TITLE
Bug in DependencyMatcher--success of a match depends on history of how patterns are added.

### ISSUE THREAD

**Issue by mdc713** [explosion/spaCy#9263]
This is either a bug or some very unexpected behavior. The results of DependencyMatcher seem to depend on the how exactly the rules are added--in what order, with what names, etc... Examples below.

## How to reproduce the behaviour
The initial set up is this.
```
from spacy.matcher import DependencyMatcher
import spacy

LID = "LEFT_ID"
RID = "RIGHT_ID"
RATTRS = "RIGHT_ATTRS"
RELOP = "REL_OP"
ORTH = "ORTH"
REGEX = "REGEX"
DEP = "DEP"

has_have = {RID: "have/has", RATTRS: {ORTH: {REGEX: "have|has|is"}}}
nsubj = {LID: "have/has", RELOP: ">", RID: "nsubj", RATTRS: {DEP: "nsubj"}}
pobj = {LID: "have/has", RELOP: ">>", RID: "pobj", RATTRS: {DEP: "pobj"}}
dobj = {LID: "have/has", RELOP: ">", RID: "dobj", RATTRS: {DEP: "dobj"}}
relcl = {LID: "have/has", RELOP: "<", RID: "relcl", RATTRS: {DEP: "relcl"}}
conj = {LID: "have/has", RELOP: ">>", RID: "conj", RATTRS: {DEP: "conj"}}

TEST_SENTENCE = "Product A and Product B have one processor, which is one more than Product C."
MODEL_NAME = "en_core_web_sm"
nlp = spacy.load(MODEL_NAME)
TEST_DOC = nlp(TEST_SENTENCE)
matcher = DependencyMatcher(nlp.vocab)
matcher.add('rule-1', [[has_have, nsubj, dobj]])
matcher.add('rule-2', [[has_have, pobj, dobj]])
matcher.add('rule-3', [[has_have, conj, dobj]])
matcher.add('rule-4', [[has_have, nsubj, relcl]])
matcher.add('rule-5', [[has_have, pobj, relcl]])
matcher.add('rule-6', [[has_have, conj, relcl]])
matcher(TEST_DOC)
```

which give the output:
```
[(507345881037203946, [5, 1, 7]),
 (13408152019079346362, [5, 15, 7]),
 (14359486885833749881, [5, 4, 7])]
```
Now things get weird. I try to add a new rule:
```
matcher.add("test_rl", [[{RID: "is", RATTRS: {ORTH: "is"}}, {LID: 'is', RID: 'relcl', RELOP: '>', RATTRS:{DEP: "dobj"}}]])
matcher(TEST_DOC)
```
This changes nothing. But now do the following:
```
matcher.remove('test_rl')
matcher.add("test_rl", [[{RID: "is", RATTRS: {ORTH: "is"}}, {LID: 'is', RID: 'relcl', RELOP: '>', RATTRS:{DEP: "relcl"}}]])
matcher(TEST_DOC)
matcher.remove('test_rl')
matcher.add("test_rl", [[{RID: "is", RATTRS: {ORTH: "is"}}, {LID: 'is', RID: 'relcl', RELOP: '<', RATTRS:{DEP: "relcl"}}]])
matcher(TEST_DOC)
```
and the output becomes
```
[(507345881037203946, [5, 1, 7]),
 (13408152019079346362, [5, 15, 7]),
 (14359486885833749881, [5, 4, 7]),
 (439699386328600021, [10, 7])]
```
It gets stranger, if I go back to the original version, and simply re-use the same name for the first six patterns, so they are all named "rule-1", I get these results:
```[(13959110402515700616, [5, 1, 7]),
 (13959110402515700616, [5, 1, 7]),
 (13959110402515700616, [5, 4, 7]),
 (13959110402515700616, [5, 9, 7]),
 (13959110402515700616, [5, 15, 7]),
 (13959110402515700616, [5, 1, 7]),
 (13959110402515700616, [5, 4, 7]),
 (13959110402515700616, [5, 9, 7]),
 (13959110402515700616, [5, 15, 7]),
 (13959110402515700616, [10, 9, 7]),
 (13959110402515700616, [10, 9, 7]),
 (13959110402515700616, [10, 15, 7]),
 (13959110402515700616, [10, 9, 7]),
 (13959110402515700616, [10, 15, 7])]
``` 
Inspecting the individual patterns, they don't seem to be different. It seems like there is something fishy going on with how these are executed. 

## Info about spaCy
- **spaCy version:** 3.1.0
- **Platform:** macOS-10.16-x86_64-i386-64bit
- **Python version:** 3.8.11
- **Pipelines:** en_core_web_sm (3.1.0)

**Comment by polm** [comment]
That definitely sounds weird, thanks for the report! We'll take a look at it.

**Comment by polm** [comment]
Just a quick update on this - it is definitely a bug and we are working on fixing it. At least part of it is fixed by #9337. 

For the time being, I'm not sure but you may be able to get correct behavior by adding all patterns with the same label in one call to `.add()` rather than multiple calls.
